src/spawn.c

```diff
@@ -0,0 +1,1121 @@
+/*
+ *      spawn.c - this file is part of Geany, a fast and lightweight IDE
+ *
+ *      Copyright 2013 Dimitar Toshkov Zhekov <dimitar(dot)zhekov(at)gmail(dot)com>
+ *
+ *      This program is free software; you can redistribute it and/or modify
+ *      it under the terms of the GNU General Public License as published by
+ *      the Free Software Foundation; either version 2 of the License, or
+ *      (at your option) any later version.
+ *
+ *      This program is distributed in the hope that it will be useful,
+ *      but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *      GNU General Public License for more details.
+ *
+ *      You should have received a copy of the GNU General Public License along
+ *      with this program; if not, write to the Free Software Foundation, Inc.,
+ *      51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+/* An ongoing effort to improve the tool spawning situation under Windows.
+ * In particular:
+ *    - There is no g_shell_parse_argv() for windows. It's not hard to write one, 
+ *      but the command line recreated by mscvrt may be wrong.
+ *    - GLib converts the argument vector to UNICODE. For non-UTF8 arguments,
+ *      the result is often "Invalid string in argument vector at %d: %s: Invalid
+ *      byte sequence in conversion input" (YMMV). Our tools (make, grep, gcc, ...)
+ *      are "ANSI", so converting to UNICODE and then back only causes problems.
+ *    - For various reasons, GLib uses an intermediate program to start children
+ *      (see gspawn-win32.c), the result being that children's children output
+ *      (such as make -> gcc) is lost.
+ *    - With non-blocking pipes, the g_io_add_watch() callbacks are not invoked,
+ *      while with blocking pipes, the child blocks on output larger than the
+ *      pipe buffer (4KB by default), and Geany blocks waiting for the child.
+ *      (g_io_channel_set_flags(G_IO_FLAG_NONBLOCK) does not work too, but that's
+ *      a small problem)
+ *
+ * Note that if no pipe I/O with the child is required, the g_spawn...() functions
+ * will generally work. They open the child in a normal window, while spawn_...()
+ * use a minimized-by-default window, more suitable for a tool.
+ *
+ * This module does not depend on Geany.
+ */
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+
+#ifdef G_OS_WIN32
+# define _NO_W32_PSEUDO_MODIFIERS
+# include <windows.h>
+# include <errno.h>
+# include <io.h>
+#else  /* G_OS_WIN32 */
+# include <sys/types.h>
+# include <sys/wait.h>
+# include <signal.h>
+# include <unistd.h>
+# ifndef O_NONBLOCK
+# define O_NONBLOCK O_NDELAY
+# endif
+#endif  /* G_OS_WIN32 */
+
+#include "spawn.h"
+
+#ifndef _
+# define _
+#endif
+
+#ifdef G_OS_WIN32
+# define BLANKS " \t"
+#endif
+
+
+/**

```

Comment:
all this module uses doc comments `/** ... `, but we reserve this for plugin API.  Do we want to add this to the API? if we do we need to actually do so, otherwise we just have to strip the second star